docs/source/API/containers/Unordered-Map.rst

```diff
@@ -0,0 +1,147 @@
+  between memory spaces) the map uses integer indexes. This also allows the map to store data
+  in cache friendly ways. The availability of indexes is managed by an internal atomic bitset based on ``uint32_t``.
+
+
+Description
+-----------
+
+.. cppkokkos:class:: template <typename Key, typename Value, typename Device = Kokkos::DefaultExecutionSpace> UnorderedMap
+
+   :tparam Key: Must be a POD
+
+   :tparam Value: `void` indicates an unordered set, otherwise must be trivially copyable
+
+   :tparam Device: Device is any class or struct with the following public typedefs or type aliases: `execution_space`, `memory_space`, and `device_type`
+
+   .. rubric:: Constructor
+
+   .. cppkokkos:function:: UnorderedMap(uint32_t capacity_hint);
+
+      Create map with enough space for at least capacity_hint number of objects
+
+      .. warning:: Host Only
+
+   .. rubric:: Public Member Functions
+
+   .. cppkokkos:function:: clear();
+
+      Clear the map
+
+      .. warning:: Host Only
+
+   .. cppkokkos:function:: bool rehash(uint32_t requested_capacity);
+
+      Rehash map to given capacity, the current size is used as a lower bound O(capacity)
+
+      .. warning:: Host Only
+
+   .. cppkokkos:function:: uint32_t size() const;
+
+      Current size of the map, O(capacity)
+
+      .. warning:: Host Only
+
+   .. cppkokkos:kokkosinlinefunction:: uint32_t capacity() const;
+
+       Capacity of the map, O(1)
+
+   .. cppkokkos:kokkosinlinefunction:: UnorderedMapInsertResult insert(key) const;
+
+      Insert the given key into the map with a default constructed value
+
+   .. cppkokkos:kokkosinlinefunction:: UnorderedMapInsertResult insert(Key key, Value value) const;
+
+      Insert the given key/value pair into the map
+
+   .. cppkokkos:kokkosinlinefunction:: uint32_t find(Key key) const
+
+      Return the index of the key if it exist, otherwise return invalid_index
+
+   .. cppkokkos:kokkosinlinefunction:: bool exists(Key key) const;
+
+      Does the key exist in the map
+
+   .. cppkokkos:kokkosinlinefunction:: bool valid_at(uint32_t index) const;
+
+      Is the current index a valid key/value pair
+
+   .. cppkokkos:kokkosinlinefunction:: Key key_at(uint32_t index) const;
+
+      Return the current key at the index
+
+   .. cppkokkos:kokkosinlinefunction:: Value value_at(uint32_t index) const;
+
+      Return the current value at the index
+
+   .. cppkokkos:kokkosinlinefunction:: constexpr bool is_allocated() const;
+
+      Return true if the internal views (keys, values, hashmap) are allocated
+
+
+.. cppkokkos:class:: UnorderedMapInsertResult
+
+   .. rubric:: Public Methods
+
+   .. cppkokkos:kokkosinlinefunction:: bool success() const;
+
+      Was the key/value pair successfully inserted into the map
+
+   .. cppkokkos:kokkosinlinefunction:: bool existing() const;
+
+      Is the key already present in the map
+
+   .. cppkokkos:kokkosinlinefunction:: bool failed() const;
+
+      Did the insert fail?
```

Comment:
Is an error emitted upon failure, or does a changed boolean just cruise on by, and no one is the wiser?